_posts/2015-03-30-geomesa-feature-level-visibility.md

```diff
@@ -0,0 +1,168 @@
+---
+title: Feature Level Visibility and Security
+author: jake
+layout: tutorial
+---
+
+{% include tutorial-header.html %}
+
+<!-- add some style to fix the xml formatting color -->
+<style>
+code.xml { color:#93a1a1 }
+</style>
+
+## Introduction
+
+Feature visibility is a new and important feature recently supported by GeoMesa. Accumulo has support for cell-level visibility built-in, and in recent versions of GeoMesa,
+support was added to utilize these visibility markings and enhance data security.
+<!--more-->
+From the Accumulo userguide:
+
+> Accumulo extends the BigTable data model to implement a security mechanism known as cell-level security. Every key-value pair has its own security label, stored under the column visibility element of the key, which is used to determine whether a given user meets the security requirements to read the value. This enables data of various security levels to be stored within the same row, and users of varying degrees of access to query the same table, while preserving data confidentiality.
+
+In this tutorial, you'll be guided through ingesting data with varying levels of visibility and querying that data as different users through GeoServer.
+
+## Prerequistes
+
+If you haven't already [read through the Quickstart tutorial](http://www.geomesa.org/2014/05/28/geomesa-quickstart/) and
+gone through the initial setup of GeoMesa, read through that first. We'll be using a customized version of the data generated by the GeoMesa Quickstart project.
+
+## Switching to the "feature_level_visibility" branch
+
+In your geomesa-quickstart directory, checkout the branch "feature_level_visibility" and build with Maven:
+
+{% highlight bash %}
+cd geomesa-quickstart
+git pull origin master
+git checkout feature_level_visibility
+mvn clean install
+{% endhighlight %}
+
+## Understanding the code
+
+If you inspect the Quickstart code in your editor of choice, you'll notice only a few changes from the original Quickstart project. The main change is that the
+SimpleFeatureType has an additional attribute, `Visibility`, added to it. This attribute will assist later in testing that features were written with the appropriate
+Accumulo visibilities.
+
+Additionally, if you look at line 169, you'll see:
+
+{% highlight scala %}
+if (i % 2 == 0) {
+    simpleFeature.setAttribute("Visibility", "admin");
+    simpleFeature.getUserData().put(SecurityUtils.FEATURE_VISIBILITY, "admin")
+} else {
+    simpleFeature.setAttribute("Visibility", "user|admin");
+    simpleFeature.getUserData().put(SecurityUtils.FEATURE_VISIBILITY, "user|admin");
```

Comment:
might want to change these methods to use the static calls, which are a bit easier: org.locationtech.geomesa.core.security.SecurityUtils#setFeatureVisibility